### PR TITLE
Add cameras

### PR DESCRIPTION
**Pull request by KevinBasener**
* Player camera
* MainCamera, that also is able to zoom in and out
* Switch between the two cameras
* Add signals for communication between player, main and map

### CODE REVIEW

**Review by hfxbse — CHANGES_REQUESTED** [review]
Please for the love of god rebase your changes on to the `main`-branch to contain only the stuff that's relevant for the camera. I won't look at 46 files changed with 1615 lines added.

**Comment by KevinBasener** [comment]
> Please for the love of god rebase your changes on to the `main`-branch to contain only the stuff that's relevant for the camera. I won't look at 46 files changed with 1615 lines added.

Now its only 7 files

**Comment by hfxbse** [comment]
Is there a possibility to set the "zoom"-level consistently? It still looks too small on a high DPI screen and this could be fixed with the camera. In addition, you should be able to move the camera once it is not following the player anymore even with just a touch screen and without a middle click. This includes a button to switch between the camera modes without a keyboard.

Also if we like to follow the Python style recommendations there should be two blank lines before the start of a function definition.

**Comment by ubeF** [comment]
There seems to be no button for unlocking and moving the camera?

**Comment by KevinBasener** [comment]
All issues should be addressed

File: main/MainCamera.gd
Comment (hfxbse) [inline main/MainCamera.gd:1]: File name should also be in snake case as the rest.

File: main/MainCamera.gd [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+const ZOOM_INCREMENT = 0.1
+
+func _ready():
+	last_pos = position
+
+func _unhandled_input(event: InputEvent) -> void:
+	if is_current():  
+		if event is InputEventMouseMotion:
+			if event.button_mask == MOUSE_BUTTON_MASK_LEFT:
+				position -= event.relative
```
Comment (hfxbse) [inline main/MainCamera.gd]: The applied position offset should scale with the zoom level as it feels like you drag the camera over a sticky surface.

File: main/main.gd [hunk 2]
```diff
@@ -3,20 +3,41 @@ extends Node
...
+
+signal turn_start()
 
 func _init():
 	var question_handler = preload("res://data_providers/resources/question_provider.tres")
 	#Example call
 	print(question_handler.get_next_question(Question.Difficulty.EASY).question)
 	print(question_handler.get_random_question(Question.Difficulty.EASY).question)
-
+
```
Comment (hfxbse) [inline main/main.gd]: There is an unnecessary tab.

**Review by hfxbse — CHANGES_REQUESTED** [review]
Not sure who it is done but when switching to the free camera it always jumps to the game origin which is a bit weird. Imo it should start a little bit zoomed out at the current player position.

I also set the correct scaling mode to not introduce up-scaling artifacts and use the whole screen aspect ratio with 50392e8f62f7ca024ba09dd2ab3c37d74618c62c. You need to anchor the camera button to the left bottom then to be in the correct position regardless of aspect ratio.

I also don't see a real reason to not place the HUD scene in the main scene, as it is always drawn. You can still remove the additional buttons if needed by referencing the HUD element in the scene. It requires less code an is less unexpected where the buttons will be placed.

As I would classify the button as an overlay, it might be appropriate to move stuff related to it in an directory called `overlay`.

File: main/MainCamera.gd (hunk 1 above)
Comment (hfxbse) [reply main/MainCamera.gd:17]: Had a second look, apparently it already does it, but it still is not enough.

File: main/MainCamera.gd [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+func zoom_out():
+	target_zoom = max(target_zoom - ZOOM_INCREMENT, MIN_ZOOM)
+	set_physics_process(true)
+	
+func _physics_process(delta):
+	zoom = lerp(zoom, 
+	target_zoom * Vector2.ONE,
+	zoom_rate * delta
+	)
```
Comment (hfxbse) [inline main/MainCamera.gd]: The call to `lerp` should be on one line, or at least have correct indents.

**Review by hfxbse — CHANGES_REQUESTED** [review]
White spaces are free :)

**Comment by KevinBasener** [comment]
> Not sure who it is done but when switching to the free camera it always jumps to the game origin which is a bit weird. Imo it should start a little bit zoomed out at the current player position.
> 
> I also set the correct scaling mode to not introduce up-scaling artifacts and use the whole screen aspect ratio with [50392e8](https://github.com/hfxbse/mparty/commit/50392e8f62f7ca024ba09dd2ab3c37d74618c62c). You need to anchor the camera button to the left bottom then to be in the correct position regardless of aspect ratio.
> 
> I also don't see a real reason to not place the HUD scene in the main scene, as it is always drawn. You can still remove the additional buttons if needed by referencing the HUD element in the scene. It requires less code an is less unexpected where the buttons will be placed.
> 
> As I would classify the button as an overlay, it might be appropriate to move stuff related to it in an directory called `overlay`.

Should be addressed

File: main/main_camera.gd [hunk 4]
```diff
@@ -14,7 +14,7 @@ func _unhandled_input(event: InputEvent) -> void:
 	if is_current():  
 		if event is InputEventMouseMotion:
 			if event.button_mask == MOUSE_BUTTON_MASK_LEFT:
-				position -= event.relative
+				position -= event.relative * zoom
```
Comment (hfxbse) [inline main/main_camera.gd]: When zooming, in the zoom value increases and when zooming out, the zoom value decreases.
It should be `position -= event.relative / zoom`. Multiplying made it even worse.

File: main/main_camera.gd [hunk 5]
```diff
@@ -31,20 +31,11 @@ func zoom_out():
...
 	
 func _physics_process(delta):
-	zoom = lerp(zoom, 
-	target_zoom * Vector2.ONE,
-	zoom_rate * delta
-	)
+	zoom = lerp(zoom, target_zoom * Vector2.ONE, zoom_rate * delta)
 	set_physics_process(
 		not is_equal_approx(zoom.x, target_zoom)
 	)
```
Comment (hfxbse) [inline main/main_camera.gd]: `set_physics_process` might be on one line as well.

**Review by hfxbse — CHANGES_REQUESTED** [review]
b086db4c462f1554642e8958c3239e55e43f2cf6 broke the main scene dependency on the HUD scene somehow and it will not let me select the correct dependency for an reason unknown to me.

250669ec616feffbe2ca89df885e8ef2c734e064 looked good besides white spaces so I think it is nearly ready to merge.

File: main/main_camera.gd (hunk 4 above)
Comment (KevinBasener) [reply main/main_camera.gd]: Good call

File: main/main_camera.gd (hunk 5 above)
Comment (KevinBasener) [reply main/main_camera.gd]: changed it

**Comment by KevinBasener** [comment]
> [b086db4](https://github.com/hfxbse/mparty/commit/b086db4c462f1554642e8958c3239e55e43f2cf6) broke the main scene dependency on the HUD scene somehow and it will not let me select the correct dependency for an reason unknown to me.
> 
> [250669e](https://github.com/hfxbse/mparty/commit/250669ec616feffbe2ca89df885e8ef2c734e064) looked good besides white spaces so I think it is nearly ready to merge.

Hope dependency doesn't break anymore

**Review by hfxbse — APPROVED** [review]

**Pull request merged**